### PR TITLE
Set timeout on calling delegator

### PR DESCRIPTION
**Pull request by nick-bisonai**
# Description

sets timeout limit into **3 sec** for calling `sendTransactionDelegatedFee`

added testcode for `axiosTimeout`
image

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

## Deployment

- [ ] Should publish npm package
- [ ] Should publish Docker image

### CODE REVIEW

File: core/src/reporter/reporter.ts
Comment (martinkersner) [inline core/src/reporter/reporter.ts:35]: How about hiding then name of error from here and define error in `OraklErrorCode` like `DelegatorTimeoutError`. Additionally, I think it would be better to make it `else if` rather than just `if` to make it more readable. In `core`, we do not use `console`, but Pino logger which is imported to `reporter` function as `logger`.

```typescript
  if (e.code == OraklErrorCode.DelegatorServerIssue) {
    delegatorOkay = false
  }
  else if (e.code == OraklErrorCode.DelegatorTimeoutError) {
    logger.error('Delegator post request timeout')
  }
```

File: core/src/reporter/utils.ts
Comment (martinkersner) [inline core/src/reporter/utils.ts:188]: How about making the timeout updatable through environment variable defined at `settings.ts`?

**Review by martinkersner — CHANGES_REQUESTED** [review]
I think we should make a couple changes before we merge it. Please check comments below.

**Comment by nick-bisonai** [comment]
@martinkersner I have updated the code, please review once more. Thank you

**Review by bayram98 — APPROVED** [review]
LGTM!

File: core/src/reporter/utils.ts
Comment (martinkersner) [inline core/src/reporter/utils.ts:193]: We have to detect what kind of error happened in this `catch` block and distinguish between `DelegatorTimeoutError` and other errors (`DelegatorServerIssue`).

**Review by martinkersner — CHANGES_REQUESTED** [review]
It would be much better if we can take the axios call outside of the `sendTransactionDelegatedFee` and write tests for it. That way we could catch potential implementation issues sooner 🙌 This is not a critical issue so we can spend more time on it and make it better.

**Comment by nick-bisonai** [comment]
@martinkersner 
I have made some updates based on your feedbacks, I'm not sure if I understood you correctly, please leave more feedbacks if I'm wrong. thank you

**Comment by nick-bisonai** [comment]
further works to be done
- extract axios call into seperate function
- add test code with different options for axios call (`get`, `post`, `put`...)

File: core/src/utils.ts
Comment (martinkersner) [inline core/src/utils.ts:158]: It would be nice if this function can be used for any axios calls, but just the ones, where we want to define timeout. Timeout should be optional same as it is in axios. What do you think?

**Review by martinkersner — COMMENTED** [review]
I left few comments. Please check them and let me know what you think 🙌

File: core/src/utils.ts
Comment (nick-bisonai) [reply core/src/utils.ts:158]: Do you mean we should add timeout parameter for all url requests? such as `axiosTimeout.get(url, data, timeout)`?

**Comment by nick-bisonai** [comment]
@martinkersner 
Updated based on the feedback ;)

File: core/src/reporter/utils.ts
Comment (martinkersner) [inline core/src/reporter/utils.ts:183]: @nick-bisonai I think we should convert the `DELEGATOR_TIMEOUT` inside of settings.

```suggestion
      await axios.post(endpoint, { ...transactionData }, { timeout: DELEGATOR_TIMEOUT })
```


settings.ts

```
export const DELEGATOR_TIMEOUT = Number(process.env.DELEGATOR_TIMEOUT) || 3000
```

**Review by martinkersner — APPROVED** [review]
LGTM! If we want to be pedantic, the name `getAxiosErrorCode` is not precise name, because we convert axios error code to orakl error code 🤔

**Comment by nick-bisonai** [comment]
> LGTM! If we want to be pedantic, the name `getAxiosErrorCode` is not precise name, because we convert axios error code to orakl error code 🤔

how about `wrapAxiosErrorCode`? or `convertAxiosErrorCode`?

**Comment by martinkersner** [comment]
> > LGTM! If we want to be pedantic, the name `getAxiosErrorCode` is not precise name, because we convert axios error code to orakl error code 🤔
> 
> how about `wrapAxiosErrorCode`? or `convertAxiosErrorCode`?

How about `getOraklErrorCode`?

**Comment by nick-bisonai** [comment]
> > > LGTM! If we want to be pedantic, the name `getAxiosErrorCode` is not precise name, because we convert axios error code to orakl error code 🤔
> > 
> > 
> > how about `wrapAxiosErrorCode`? or `convertAxiosErrorCode`?
> 
> How about `getOraklErrorCode`?

is this naming assuming that the function will be extended for further use later on? not only for axios? If that's the case I agree

**Comment by martinkersner** [comment]
> > > > LGTM! If we want to be pedantic, the name `getAxiosErrorCode` is not precise name, because we convert axios error code to orakl error code 🤔
> > > 
> > > 
> > > how about `wrapAxiosErrorCode`? or `convertAxiosErrorCode`?
> > 
> > 
> > How about `getOraklErrorCode`?
> 
> is this naming assuming that the function will be extended for further use later on? not only for axios? If that's the case I agree

There can be other errors returned by Orakl Network Delegator server. One of which we have already set as a default error code. In such cases I think we do not just `wrap` or `convert` axios code 🤔 But I am also fine with your suggestions 🙌

**Pull request merged**
